### ISSUE TITLE
Avoid forced evaluation of the result returned by a function

### ISSUE THREAD

**Issue by Necr0x0Der** [trueagi-io/hyperon-experimental#500]
The problem is that it is impossible to stop evaluation of the result returned by some function. For example, we want to print the result of `get-atoms`. If we quote `get-atoms` (wrap into a function that accepts `Atom`), it will not be evaluated at all. If we call it even inside `let` for quoting after this, its result will be further reduced, e.g., `(let $content (get-atoms &self) (quote $content))`.
The same problem arises for lazy evaluations. While lazy `(: if (-> Bool Atom Atom $t))` works, it is not enough to write, say, lazy `all`. This will not work:
```
(: all (-> Atom Bool))
(= (all $A)
   (if (== () $A) True
       (if (car-atom $A) (let $cdr (cdr-atom $A) (all $cdr)) False))
)
(= (loop) (loop))
!(all (True False (loop) True))
```
because once `(cdr-atom $A)` is called, its result `(False (loop) True))` will immediately be evaluated further.
Possible solutions:
* [Originally proposed] If the function return type is `Atom`, the interpreter doesn't evaluate it further.
* [Hacky and inconvenient, but simple] A special grounded function is introduced, which executes the given grounded function (without full interpretation) and quotes it. Say, `(step (cdr-atom (True False (loop) True)))` will return `(quote (False (loop) True))`.
* A solution in Minimal MeTTa?

**Comment by vsbogd** [comment]
> A solution in Minimal MeTTa?

Yes, minimal MeTTa has such solution, using `chain` + `return` allows returning atom from function and inserting it into a template without further evaluation.

**Comment by Necr0x0Der** [comment]
It's slightly tricky, but now it can work like this:

```
(: q (-> Atom Atom))
(= (q $x) $x)

(: all (-> Atom Bool))
(= (all $A)
   (if (== () (q $A)) True
       (if (car-atom $A) (let $cdr (cdr-atom $A) (all $cdr)) False))
)
(= (loop) (loop))
!(all (True False (loop) True))
```

which makes me think that it could work earlier with
```
(if (== (quote ()) (quote $A)) True ...)
```

So, this comparison of `$A` was the root cause of `(loop)` inside `$A` to be evaluated. But maybe it wouldn't be enough earlier, because `(cdr-atom $A)` would still be evaluated (now it is not because `cdr-atom` return type is `Expression`).
In any case, it works either way, `get-atoms` also works now, so I'm closing it.
